README.md
# networking
A network packet sniffer.

Netdump is a Network Packer Sniffer, which will decode and print out packets being transported over the network.

pingStuff was some code I threw together to do a homework assignment for me.


netdump/changelog.txt
6/27
  -Got TCP decoded. 

6/24
  -Finally figured out bitshift
  -Another bug, I switched the decoding for offset & flag
  -Found first bug in IP. I am doing the two byte thing wrong. I was
     for ex:,
        uint16_t length = p[4] * 256 + p[5]
	uint16_t id = p[5] *256 + p[6];
     I was overlapping on the numbers.

6/23
    -ARP looks fine, but there is still something wrong with IP.
       I can't get any ICMP packets, my flags are always 000 and
       my protocol is always 0. So I am doing something wrong.

6/22
    -Got tired of the numbers, started doing indexes on the packet incrementer
6/16
    -Have lot of stuff commented out to just work on ARP packets. Will finish
       ARP tomorrow
    -Going to start testing my ARP decode/print. will need to refactor, becuase
    	it looks pretty poorly done.
    -Trying to figure out teh variable length on ARP. It says when it is Ethernet
    	then the value is 6. Does this mean that Hardware address is depdent on hardware
	type? Well when I write it, it seems obvious. But I don't see anywhere in the
	text that explicity says that.
    -Decoding ARP and the IP header. But in those methods I have p[14] and on
        Also I am just decoding and printing on one line. I should seperate this

6/15
    -Looks like 5.1a b, c, and d are all totally done. 
    -Printing in hex, also printing the type/length fields
    -Adding debugging symbols to program to see how it works.
    -syntax error when introducing counting packets & displaying
       Ethernet type






/Misc dates
PA1

   -Changed packettype to packet_type


netdump/packet.h
#ifndef PACKET_H
# define PACKET_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

/*CONSTANTS */
#define ADDR_LENGTH 6
#define ARP_REQUEST 1
#define ARP_REPLY 2
/* Used to keep information about the current Session*/
typedef struct{
  int broadcast_packets_total;
  int ip_packets_total;
  int arp_packets_total;
  int icmp_packets_total;  
  int tcp_packets_total;
  int udp_packets_total;
}Session;

extern Session currSession;

typedef struct{
  u_char *raw_data;
  u_int length;
  u_int caplen;
  u_char dest_addr[ADDR_LENGTH];
  u_char src_addr[ADDR_LENGTH];
  uint16_t type_length;
}Packet;

typedef struct{
  Packet info;
  uint16_t hw_type;
  uint16_t protocol_type;
  uint8_t hw_len;
  uint8_t protocol_len;
  uint16_t operation; //if 1 then 
}ARP;

void print_packet_header(const u_char *packet);
void decode_ARP_packet(const u_char *packet_data);//6a
void decode_IP_header(const u_char *packet);//6b
void decode_ICMP_header(const u_char *packet);//6c
void decode_TCP_header(const u_char *packet);//7a
#endif


netdump/netdump.c
/*
 * Packet Sniffer 
 * ---------------
 */
#define RETSIGTYPE void

#include <sys/types.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <pcap.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "packet.h"
#ifndef setsignal_h
#define setsignal_h

RETSIGTYPE (*setsignal(int, RETSIGTYPE (*)(int)))(int);
#endif

char cpre580f98[] = "netdump";//not sure yet

void raw_print(u_char *user, const struct pcap_pkthdr *h, const u_char *p);

int packet_type; //ICMP(1), UDP(17), TCP(6) etc.

char *program_name;

/* Externs */
extern void bpf_dump(const struct bpf_program *, int); //Berkey packet Filter

extern char *copy_argv(char **);

// Forwards //Edited THis, added the session struct
void program_ending(int);

/* Length of saved portion of packet. */
int snaplen = 1500;

static pcap_t *pd;

extern int optind;
extern int opterr;
extern char *optarg;
int pflag = 0, aflag = 0;
Session currSession;
int main(int argc, char **argv) {
	int cnt, op, i, done = 0;
	bpf_u_int32 localnet, netmask;
	char *cp, *cmdbuf, *device;
	struct bpf_program fcode;
	void (*oldhandler)(int);
	u_char *pcap_userdata;
	char ebuf[PCAP_ERRBUF_SIZE];

	int run_ad_infinitum = -1;
	cnt = run_ad_infinitum;
	device = NULL;
	currSession.broadcast_packets_total = 0;
	currSession.ip_packets_total = 0;
	currSession.arp_packets_total = 0;
	currSession.icmp_packets_total = 0;
	currSession.tcp_packets_total = 0;
	currSession.udp_packets_total = 0;
	if ((cp = strrchr(argv[0], '/')) != NULL)
		program_name = cp + 1;
	else
		program_name = argv[0];

	opterr = 0;
	while ((i = getopt(argc, argv, "pa")) != -1) {
	  switch (i) {
	  case 'p':
	    pflag = 1;
	    break;
	  case 'a':
	    aflag = 1;
	    break;
	  case '?':
	  default:
	    done = 1;
	    break;
	  }
	  if (done){
	    break;
	  }
	}
	if (argc > (optind)){
	  cmdbuf = copy_argv(&argv[optind]);
	} 
	else {
	  cmdbuf = "";
	}

	if (device == NULL) {
	  device = pcap_lookupdev(ebuf);
	  if (device == NULL){
	    error("%s", ebuf);
	  }
	}

	pd = pcap_open_live(device, snaplen,  1, 1000, ebuf);
	if (pd == NULL){
	  error("%s", ebuf);
	}
	i = pcap_snapshot(pd);
	if (snaplen < i) {
	  warning("snaplen raised from %d to %d", snaplen, i);
	  snaplen = i;
	}
	if (pcap_lookupnet(device, &localnet, &netmask, ebuf) < 0) {
	  localnet = 0;
	  netmask = 0;
	  warning("%s", ebuf);
	}
	/*
	 * Let user own process after socket has been opened.
	 */
	setuid(getuid());
	
	if (pcap_compile(pd, &fcode, cmdbuf, 1, netmask) < 0){
	  error("%s", pcap_geterr(pd));
	}
	
	(void)setsignal(SIGTERM, program_ending);
	(void)setsignal(SIGINT, program_ending);
	/* Cooperate with nohup(1) */
	if ((oldhandler = setsignal(SIGHUP, program_ending)) != SIG_DFL){
	  (void)setsignal(SIGHUP, oldhandler);
	}
	if (pcap_setfilter(pd, &fcode) < 0){
	  error("%s", pcap_geterr(pd));
	}
	pcap_userdata = 0;
	(void)fprintf(stderr, "%s: listening on %s\n", program_name, device);

	/*pd = the packet
	* cnt = number of packets to run (-1 means infinity)
	* raw_print = pcap handler callback
	* pcap_userdata = 
	*/
	if (pcap_loop(pd, cnt, raw_print, pcap_userdata) < 0) {
	  (void)fprintf(stderr, "%s: pcap_loop: %s\n",
			program_name, pcap_geterr(pd));
	  exit(1);
	}
	pcap_close(pd);
	exit(0);
}

/*Programin is ending,  routine is executed on exit */
void program_ending(int signo) {
	struct pcap_stat stat;
	
	if (pd != NULL && pcap_file(pd) == NULL) {
	  (void)fflush(stdout);
	  putc('\n', stderr);
	  if (pcap_stats(pd, &stat) < 0){
	    (void)fprintf(stderr, "pcap_stats: %s\n", pcap_geterr(pd));
	  }
	  else {
	    (void)fprintf(stderr, "%d packets received by filter\n",
			  stat.ps_recv);
	    (void)fprintf(stderr, "%d packets dropped by kernel\n",
			  stat.ps_drop);
	    
	    (void)fprintf(stderr, "Number of Broadcast Packets = %d\n",
			  currSession.broadcast_packets_total);
	    (void)fprintf(stderr, "Number of IP Packets = %d\n",
			  currSession.ip_packets_total);
	    (void)fprintf(stderr, "Number of ARP Packets = %d\n",
			  currSession.arp_packets_total);
	    (void)fprintf(stderr, "Number of ICMP Packets = %d\n",
			  currSession.icmp_packets_total);
	    (void)fprintf(stderr, "Number of TCP Packets = %d\n",
			  currSession.tcp_packets_total);    	    
	  }
	}
	exit(0);
}

/* Like default_print() but data need not be aligned */
void default_print_unaligned(register const u_char *cp, register u_int length) { 
  register u_int i, s; //register means it will be used a lot, it is a tip to the compiler
	register int nshorts;

	nshorts = (u_int) length / sizeof(u_short);
	i = 0;
	while (--nshorts >= 0) {
	  if ((i++ % 8) == 0){
	    (void)printf("\n\t\t\t");
	  }
	  s = *cp++;
	  (void)printf (" %02x%02x ", s, *cp++);// This is printing two sets of info. s will jump another spot, and this will print the next two things	
	}
	if (length & 1) {
	  if ((i % 8) == 0){
	    (void)printf("\n\t\t\t");
	  }
	  (void)printf(" %02x", *cp);
	}
}

/*
 * By default, print the packet out in hex.
 */
void default_print(register const u_char *bp, register u_int length) {
	register const u_short *sp;
	register u_int i;
	register int nshorts;

	if ((long)bp & 1) {
		default_print_unaligned(bp, length);
		return;
	}
	sp = (u_short *)bp;
	nshorts = (u_int) length / sizeof(u_short);
	i = 0;
	while (--nshorts >= 0) {
	  if ((i++ % 8) == 0){
	    (void)printf("\n\t");
	  }
	  (void)printf(" %04x", ntohs(*sp++));
	}
	if (length & 1) {
	  if ((i % 8) == 0){
		(void)printf("\n\t");
	  }
	  (void)printf(" %02x", *(u_char *)sp);
	}
}

/*
 * This is used as the callback function for pcap_loop. 
 * it is called every time a packet is received. 
 * caplen is the length of the ethernet packet
 * character array p, which is the packet p[0] DEST hardware address
 */
void raw_print(u_char *user, const struct pcap_pkthdr *h, const u_char *p) {
  u_int length = h->len;
  u_int caplen = h->caplen; //the length of the ethernet packet

  printf("\n\t -----------[START OF DECODE]------------\n");  
  print_packet_header(p);
  printf("\n\t -------[END OF DECODE]-------\n");
  printf("\n\t -----------[START OF RAW DATA]-----------\n");
  default_print(p, caplen);
  putchar('\n');  
  printf("\n\t ------------[END OF RAW DATA]------------\n");
  //  free(currPacket.raw_data);
}




pingStuff/parsePing.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <dirent.h>
#include <unistd.h>

int pos(char *line, const char *target, int offset);
void get_avg_in_file(char *filepath);
char *filepath_concat(const char *home, const char *directory);
char *extract_substring(char * line, const char *begin,
			const char *end, int offset);

int main(int argc, char *argv[]){
  if(argc == 1){
    printf("No file given\n");
    return -1;
  }

  //run everything in this directory
  if(argv[1][0] == '-' && argv[1][1] == 'd' && argc > 2){
    //get file path
    //char *home = getenv("HOME");
    char *given = argv[2];
    
    //get all the contents of the directory
    struct dirent *directory;
    DIR *dr = opendir(given);
    if(dr == NULL){
      printf("Could not open: %s\n", given);
      //free(filePath);
      return -404;
    }
    while((directory = readdir(dr)) != NULL){
      char *currFileName = directory->d_name;
      if(currFileName[0] != '.'){
	//printf("File: %s\n", directory->d_name);
	
	get_avg_in_file(currFileName);
      }

    }
    
    closedir(dr);
    //free(filePath);
    return 0;
  }

  /*
  FILE *fp;
  static char buffer[1024];
  
  fp=fopen(argv[1], "r");
  
  int numOfLines = 0;
  float runningTotal = 0;

  static const char *target = "time=";

  while(fgets(buffer, 1024, fp)){
    //Get the Section I want
    //printf("%s", line);
    numOfLines++;
    int timePos  = pos(buffer, target, 4);
    char section[11];
    memmove(section, &buffer[timePos], 11);
    section[11] ='\0';
    //    printf("7th in List: %s\n", hope);
    
    //Extract the value out
    int valuePart = pos(section, "=", 1) + 1;
    char part2[6];
    memmove(part2, &section[valuePart], 5);
    part2[6] = '\0';
    //printf("Is this the number?: %s\n", part2);

    //convert string value into float
    float currValue = atof(part2);
    runningTotal+= currValue;
  }
  fclose(fp);
  float avg = -1;
  float tempLines = numOfLines;
  avg = runningTotal / tempLines;
  printf("Ran file: %s\n", argv[1]);
  printf("Running total was %f\n", runningTotal);
  printf("Total input of %d, and average value of: %0.3f\n", numOfLines, avg);
  printf("==================================================\n\n");
  */
  
  return 0; 
}

int pos(char *line, const char *target, int offset){
  char input[strlen(line)];
  memcpy(input, line+offset, strlen(line)-offset + 1);
  char *p = strstr(input, target);
  if(p){
    return p - input+offset;
  }
  return -1;
}

int last_index_of(char *line, const char target){
  int i;
  for(i = strlen(line) - 1; i > -1; i--){
    if(target == line[i]){
      return i;
    }
  }
  return -2;
}

char *extract_substring(char * line, const char *begin,
			const char *end, int offset){
  char input[strlen(line)];
  memcpy(input, line+offset, strlen(line - offset + 1));
  return NULL;
}

char *filepath_concat(const char *directory, const char *fileName){
  //currently the input is cwd (47) + file name + /
  char *s0 = malloc(strlen(directory) + strlen(fileName) + 1); //+1 for nullbit
  strcpy(s0, directory);
  strcat(s0, fileName);
  return s0;
}

//Need the address for this thing.
void get_avg_in_file(char *filePath){
  //for now
  char *front = "/home/will/summer19/networking/pingStuff/data/";
  FILE *fp;
  
  static char buffer[1024];
  char *path = filepath_concat(front, filePath);
  
  fp=fopen(path, "r");
  
  int numOfLines = 0;
  float runningTotal = 0;

  static const char *target = "time=";

  while(fgets(buffer, 1024, fp)){
    //Get the Section I want
    //printf("%s", line);
    numOfLines++;
    if(numOfLines == 100){
      break;
    }
    int timePos  = pos(buffer, target, 4);
    int endPos = last_index_of(buffer, ' ');
    if(timePos != -1){
      int length = endPos - timePos;
      char section[length + 1];//for null byte
      memmove(section, &buffer[timePos], length);
      section[length + 1] ='\0';
      //    printf("7th in List: %s\n", hope);
      
      //Extract the value out
      int valuePart = pos(section, "=", 1) + 1;
      int newLength = length - valuePart;
      char part2[newLength + 1]; //removing the 'time='.... techdebt right here baby
      memmove(part2, &section[valuePart], newLength);
      part2[newLength + 1] = '\0';
      //printf("Is this the number?: %s\n", part2);
      //convert string value into float
      float currValue = atof(part2);
      runningTotal+= currValue; 
    }
  }
  fclose(fp);
  float avg = -1;
  float tempLines = numOfLines;
  avg = runningTotal / tempLines;
  printf("Ran file: %s\n", path);
  printf("Running total was %f\n", runningTotal);
  printf("Total input of %d, and average value of: %0.3f\n", numOfLines, avg);
  printf("==================================================\n\n");

  free(path);
}


netdump/packet.c
#include "packet.h"

void set_header(Packet *header){}

//This should be called decode. Section 5.2 Pages 98-100.
void print_packet_header(const u_char *packet){
  currSession.broadcast_packets_total++;                                                              
  char *PACKET_PRINT_FORMAT_IPV6 = "%s = %02x:%02x:%02x:%02x:%02x:%02x\n";                          
  int i = 0;                                                                                        
  /*Print the Addresses */                                                                          
  char *dst_addr = "DEST Address";//0-5                                       
  char *src_addr = "SRC Address ";//6-10
  printf(PACKET_PRINT_FORMAT_IPV6, dst_addr, packet[i++], packet[i++],
	 packet[i++], packet[i++], packet[i++], packet[i++]);                                       
  printf(PACKET_PRINT_FORMAT_IPV6, src_addr, packet[i++], packet[i++],
	 packet[i++], packet[i++], packet[i++], packet[i++]);                                       
                                                                                                  
  /*Print the Type/Length field */                                                                  
  //If the Type/Length field is at least 1536 (0x600) then it is a protocol type.                   
  //Otherwise it is a length                                                                        
  uint16_t type_or_length = ((packet[i++] << 8) +  packet[i++]);//Or 12 
  uint16_t cut_off = 0x0600;                                                                        
  if(type_or_length < cut_off){                                                                     
    printf("Length = %0d\n", packet[12]);                                                           
  }else{//Protocol Type                                                                             
    printf("Type = 0x%04x, ", type_or_length);                                                      
           
    uint16_t IP = 0x800;                                                                            
    uint16_t ARP = 0x806;                                                                           
    uint16_t IPv6 = 0x86DD; 
    if(type_or_length == IP){                                                                       
      printf("Payload = IP\n");                                                                     
      currSession.ip_packets_total++;
      decode_IP_header(packet + 14);
    }
    else if(type_or_length == ARP){
      printf("Payload = ARP\n");                                                                    
      currSession.arp_packets_total++; 
      decode_ARP_packet(packet);                                 
      //Call function that prints out ARP                                                           
    }                                                              
    else if(type_or_length == IPv6 ){                                                               
      printf("Payload = IPv6\n");                                                                   
      //increment this?                                                                             
    }else{                                                        
      //still don't know what 0x9000 is                                                             
      //0x7bda,                                                                                     
      //0x2715,                                                                                     
      //0x1baa,                                                                                     
      //0x1856,                                                                                     
    }                                                                                               
  }     
}

//might run with sudo ./netdump arp -a
void decode_ARP_packet(const u_char *packet_data){
  printf("Arp Packet");
  int i = 14;                                                                                       
                                                                                                    
  uint16_t hw_type = packet_data[i++] * 256 + packet_data[i++];                                     
  uint16_t protocol_type = packet_data[i++] *256 + packet_data[i++];
  uint8_t hw_len = packet_data[i++];                                                                
  uint8_t protocol_len = packet_data[i++];                                                          
  uint16_t operation = packet_data[i++] * 256 + packet_data[i++];                                   
               
  u_char sender_hw_addr[hw_len];                                                                    
  int h = 0;                                                                                        
  for(h = 0; h < hw_len; h++){                                          
    sender_hw_addr[h] = packet_data[i++];                                                           
  }                                                                                                 
  //if IPv4, then it is 4 bytes, so IPv6 is 6 bytes?                                                
  uint8_t sender_protocol_addr[protocol_len];                                                       
  int p = 0;                                                                                        
  for(p = 0; p < protocol_len; p++){                                                                
    sender_protocol_addr[p] = packet_data[i++];                                                     
  }                                                                                                 
  u_char target_hw_addr[hw_len];                                                                    
  for(h = 0; h < hw_len; h++){                                                                      
    target_hw_addr[h] = packet_data[i++];                                                           
  }                                                                                                 
  uint8_t target_protocol_addr[protocol_len];                                                       
  for(p = 0; p < protocol_len; p++){                                                                
    target_protocol_addr[p] = packet_data[i++];                                                     
  }                                                                                                 
                                                                                                    
  printf("Hardware type: %u\n", hw_type);                                                           
  printf("Protocol Type: %u\n", protocol_type);                                                     
  printf("Hardware Length: %d\n", hw_len);                                                            
  printf("Protocol Length: %d\n", protocol_len);                                                    
  printf("Operation: %u\n", operation);                                                             
  if(operation == 1){                                                                               
    printf("ARP Request\n");                                                                        
  }                                                                                                 
  else if(operation == 2){                                                                          
    printf("ARP Reply\n");                                                                          
  }else {                                                                                           
    printf("Error : Unknown Arp Operation");                                                        
  }                                                                                                 
  //This is where things can get weird. Got to print out all the things with                        
  //  variables lengths. This would probably be better to put into a method                         
  //IPv4 does xxx.xxx.xxx.xxx. || IPv6 does xx:xx:xx:xx:                                            
  //Hardware addresses use the :, but I will diff between IPv later.                                
  printf("Sender Hardware Address: ");                                                              
  for(h = 0; h < hw_len; h++){//There has to be a better way to do this                             
    printf("%02x", sender_hw_addr[h]);                                                              
    printf((h + 1 < hw_len)? ":" : "\n");                                                           
  }                                   
  printf("Sender Protocol Address: ");//TODO: Look in printing diff of IPv4/6                       
  for(p = 0; p < protocol_len; p++){                                                                
    printf("%d", sender_protocol_addr[p]);                                                          
    printf((p + 1 < protocol_len)? "." : "\n");                                                     
  }                                                                                                 
  printf("Target Hardware Address: ");                                                              
  for(h = 0; h < hw_len; h++){                                                                      
    printf("%02x", target_hw_addr[h]);                                                              
    printf((h + 1 < hw_len)? ":" : "\n");                                                           
  }                                                                                                 
  printf("Target Protocol Address: ");                                                              
  for(p = 0; p < protocol_len; p++){                                                                
    printf("%d", target_protocol_addr[p]);//TODO: Look in printing diff of IPv4/6                   
    printf((p + 1 < protocol_len)? "." : "\n");                                                     
  }                                                                                                 
  
}

//"Remove" first 14 bits. Fixed Header that is 20 bytes long
void decode_IP_header(const u_char *packet){   
  printf("IP Packet Header::\n");
  //version 4 bits: IPv(4/6).
  uint8_t version;
  version = ((packet[0] & 0xF0) >> 4); //first part should negate the last 4, then remove them
  printf("Version: %u\n", version); 
  //Header length 4 bits: 4-byte words (default is 5)
  uint8_t header_len;
  header_len = ((packet[0] & 0x0F) * 4);
  printf("Header length:%u\n", header_len);

  uint8_t service_type;
  service_type = packet[1];
  printf("Service Type: %u \n", service_type);//Usually this is all 0
  uint16_t length;
  length = packet[2] * 256 + packet[3];
  printf("Length of the Payload: %u\n", length);
  //Identifier 16 bits: unique id each one. used for reassembley
  uint16_t identifier;
  identifier = packet[4] * 256 + packet[5];
  printf("Identifier: %u\n", identifier);
  //Flags 3 bits. first is reserved & set to 0.                                                     
  //     D = 1 = don't frag                         
  //     M =1, more data, =0, last packet
  uint8_t flag;//3 freaking bits. [x][y][z]
  //offset(13bits): indicate where the frag should be placed in reassembly buffer
  flag = (packet[6] >>  5); //Really gottta double check this stuff
  printf("Flag: %u %u %u\n", (flag >> 2), ((flag >> 1) & 0x01), (flag  & 0x01));
  uint16_t offset;//13 bits, number of #8 bytes
  //make room for p[7]. Then remvoe the first 3 bits. THEN add p7.
  offset = (((packet[6] << 8) & 0x1F) + packet[7]);
  printf("Offset = %u\n", offset);
  uint8_t TTL;
  TTL = packet[8];
  printf("TTL: %u\n", TTL);

  //Gateway (3), Stream(5), Exterior Gateway(8), private interrori gateay(9)
  //Network voice (11), Host Monitoring (20), Reliable (27), 22, 28, 30, 61
  uint8_t ICMP = 1;
  uint8_t TCP = 6;
  uint8_t UDP = 17;

  uint8_t protocol;
  protocol = packet[9];
  printf("Protocol: %u\n", protocol);
  
  //checksum 16bits: used for err detections
  uint16_t checksum;
  checksum = packet[10] * 256 + packet[11];
  printf("Checksum: %u\n", checksum);
  printf("Src IP Address %u.%u.%u.%u\n", packet[12], packet[13], packet[14], packet[15]);
  printf("Dst IP Address %u.%u.%u.%u\n", packet[16], packet[17], packet[18], packet[19]);

  if(protocol == ICMP){
    currSession.icmp_packets_total++;
    decode_ICMP_header(packet + header_len);
  }
  else if(protocol == TCP){
    printf("Protocol is TCP\n");
    currSession.tcp_packets_total++;
    decode_TCP_header(packet + header_len);
  }
  else if(protocol == UDP){
    printf("Protocol is UDP\n");
    currSession.udp_packets_total++;
  }else{
    printf("Not yet mapped Protocol number %u\n", protocol);
  }
}

  //only give this packet the portion it needs
void decode_ICMP_header(const u_char *packet){
  //TODO:
  printf("Protocol: ICMP::\n");
  //starts after the IP header.                                                                     
  uint8_t type = packet[0];
  uint8_t code = packet[1];
  uint16_t checksum = packet[2] * 256 + packet[3];
  uint32_t original_timestamp = ((
				  ((
				   ((
				     packet[4] << 8) + packet[5])
				   << 8) + packet[6])
				  << 8) + packet[7]);
  uint32_t receive_timestamp = ((
				 ((
				   ((
				     packet[8] << 8) + packet[9])
				   << 8) + packet[10])
				 << 8) + packet[11]);
  uint32_t transmit_timestamp = ((
				  ((
				    ((
				      packet[12] << 8) + packet[13])
				    << 8) + packet[14])
				  << 8) + packet[15]);
  printf("Type: %u, Code: %u\n", type, code);
  printf("Checksum: %u\n", checksum);
  
  if(type == 0 && code == 0){
    printf("\tICMP ECHO Reply\n");
  }
  else if(type == 8 && code == 0){
    printf("\tICMP ECHO Request\n");
  }
  else if(type == 13 && code == 0){
    printf("\tTimestamp request\n");
  }
  else if(type == 14 && code ==0){
    printf("\tTimestamp reply\n");
  }
  else if(type == 3 && (code > 0 && code <= 15)){
    printf("\tDest unreachable\n");
  }
  else if(type == 11 && code <= 1){
    printf("\tTime exeeced\n");
  }
  else if(type == 5 & code <= 3){
    printf("Redirection\n");
  }else{
    printf("Don't know yet\n");
  }

  //Time stuff only prints here
  if(type == 13 || type == 14){
    printf("Original timestamp:   %u\n", original_timestamp);
    printf("Receive timestamp:    %u\n", receive_timestamp);
    printf("Transmit timestamp:   %u\n", transmit_timestamp); 
  }                      
}

void decode_TCP_header(const u_char *packet){

  uint16_t src_port_num;
  src_port_num = ((packet[0] << 8) + packet[1]);
  printf("Source Port Num: %u\n", src_port_num);
  uint16_t dst_port_num;
  dst_port_num = ((packet[2] << 8) + packet[3]);
  printf("Destination Porn Num: %u\n", dst_port_num);
  uint32_t seq_num;
  seq_num = ((
	      ((
		((
		  packet[4] << 8) + packet[5])
		<< 8) + packet[6])
	      << 8) + packet[7]);
  printf("Sequence Num: %u\n", seq_num);
  uint32_t ack_num;
  ack_num = ((
	      ((
		((
		  packet[8] << 8) + packet[9])
		<< 8) + packet[10])
	      << 8) + packet[11]);
  printf("Acknowledge Num: %u\n", ack_num);
  uint8_t hdr_len;//4 bits
  hdr_len = (packet[12] & 0xF0) * 4; // times 4?
  printf("Header Length: %u\n", hdr_len);
  uint8_t reserved;//6 bits
  reserved = ((packet[12] & 0x3F) >> 2);
  printf("Reserved: %u\n", reserved);
  uint8_t flags;//6
  flags = (packet[13] & 0x3F);
  printf("Flags: %u\n", flags);
  printf("Flag\t          Function          \t Value \n");
  printf("URG \t Packet Contains Urgent Data \t   %u\n", flags >> 5);
  printf("ACK \t      Ack Num Is Valid  \t   %u\n", flags >> 4 & 0x01);
  printf("PSH \tData Should be pushed to Appl\t   %u\n", flags >> 3 & 0x01);
  printf("RST \t         Reset Packet       \t   %u\n", flags >> 2 & 0x01);
  printf("SYN \t      Synchronize Packet    \t   %u\n", flags >> 1 & 0x01);
  printf("FIN \t       Finish Packet        \t   %u\n", flags & 0x01);

  uint16_t window_size;
  window_size = (packet[14] << 8 + packet[15]);
  printf("Window size: %u\n", window_size);
  uint16_t checksum;
  checksum = ((packet[16] << 8) + packet[17]);
  printf("Checksum: %u\n", checksum);
  uint16_t urgent_pointer;
  urgent_pointer = ((packet[18] << 8) + packet[19]);
  printf("Urgent Pointer: %u\n", urgent_pointer);
  
  int i = 0;
  int diff = 0;
  if(hdr_len == 0){
    hdr_len = 40;
  }
  diff = hdr_len - 20;
  int offset = 20;
  printf("Options--- \n");
  for(i = 0; i < diff; i++){
    printf("%x", packet[i + offset]);
  }
  printf("\n");
  
}
